docker-base-build/test_install_pinned.py

```diff
@@ -0,0 +1,294 @@
+def test_has_exact(specifiers: str, result: bool) -> None:
+    assert install_pinned.has_exact(SpecifierSet(specifiers)) == result
+
+
+def test_parse_requirement_req() -> None:
+    req = install_pinned.parse_requirement('requests[security]>=2.8.1')
+    assert isinstance(req, install_pinned.Package)
+    assert req == Package('requests[security]>=2.8.1', constraint=False, weak=False)
+
+
+def test_parse_requirement_option() -> None:
+    req = install_pinned.parse_requirement('--no-binary pycuda')
+    assert req == '--no-binary pycuda'
+
+
+def test_parse_requirement_other() -> None:
+    with pytest.warns(UserWarning, match='Requirement foo=1 could not be parsed'):
+        req = install_pinned.parse_requirement('foo=1')
+    assert req == 'foo=1'
+
+
+@pytest.mark.parametrize(
+    'constraint, weak',
+    [(False, False), (True, False), (True, True)]
+)
+def test_parse_requirements_simple(tmp_path, constraint: bool, weak: bool) -> None:
+    filename = tmp_path / 'requirements.txt'
+    filename.write_text('''
+# A comment
+foo==0.1        # An end-of-line comment
+
+--an-option
+''')
+    reqs = install_pinned.parse_requirements(str(filename), constraint=constraint, weak=weak)
+    assert set(reqs) == {
+        Package('foo==0.1', constraint=constraint, weak=weak),
+        '--an-option'
+    }
+
+
+def test_parse_requirements_nested(tmp_path) -> None:
+    filename = tmp_path / 'base.txt'
+    (tmp_path / 'requirements.txt').write_text('required>=1.0\n')
+    (tmp_path / 'constraints.txt').write_text('constrained<=2.0\n')
+    (tmp_path / 'defaults.txt').write_text('defaulted==1.5\n')
+    filename.write_text(f'''
+-d defaults.txt
+-r requirements.txt
+-c {tmp_path / "constraints.txt"}
+''')
+    reqs = install_pinned.parse_requirements(str(filename))
+    assert set(reqs) == {
+        Package('required>=1.0', constraint=False, weak=False),
+        Package('constrained<=2.0', constraint=True, weak=False),
+        Package('defaulted==1.5', constraint=True, weak=True)
+    }
+
+
+def test_parse_requirements_http(mocker) -> None:
+    url1 = 'https://example.invalid/foo/requirements.txt'
+    url2 = 'https://example.invalid/constraints.txt'
+    content = {
+        url1:
+            b'-c ../constraints.txt\n'
+            b'foo>=1.0\n',
+        url2: b'bar==2.0\n'
+    }
+
+    def urlopen_replacement(url):
+        return io.BytesIO(content[url])
+
+    mocker.patch('urllib.request.urlopen', side_effect=urlopen_replacement)
+    reqs = install_pinned.parse_requirements(url1)
+    assert set(reqs) == {
+        Package('foo>=1.0'),
+        Package('bar==2.0', constraint=True)
+    }
+
+
+def test_merge_packages_simple() -> None:
+    pkg = install_pinned.merge_packages(
+        Package('requests [security,tests] >= 2.8.1, == 2.8.* ; python_version >= "2.7"'),
+        Package('requests [security,foo] >= 2.7')
+    )
+    assert pkg == Package(
+        'requests [security,foo,tests]>=2.7,>=2.8.1,==2.8.*; python_version >= "2.7"')
```

Comment:
Do you leave further normalisation for pip?

Reply:
Do you mean combining redundant specifiers? Yes, I just let pip figure all that out.